mdsa/algorithms/nystrom.py

```diff
@@ -0,0 +1,405 @@
+        #
+        result = zeros((nseeds, nseeds))
+        for i in range(nseeds):
+            for j in range(i, nseeds):
+                # avoid pow(x,2). it's slow
+                eij_sq = matrix_e[i, j] * matrix_e[i, j]
+                result[i, j] = -0.50 * (
+                    eij_sq -
+                    col_center[j] -
+                    row_center[i] +
+                    grand_center)
+                if i != j:
+                    result[j, i] = result[i, j]
+
+        return result
+
+    @staticmethod
+    def _build_seed_matrix(fullmat_dim, seedmat_dim, getdist, permute_order=True, PRINT_TIMINGS=False):
+        """Builds a seed matrix of shape seedmat_dim x fullmat_dim
+
+        Returns seed-matrix and indices to restore original order (needed
+        if permute_order was True)
+
+        Arguments:
+        - `fullmat_dim`:
+          dimension of the unknown (square, symmetric) "input" matrix
+        - `seedmat_dim`:
+          requested dimension of seed matrix.
+        - `getdist`:
+          distance function to compute distances. should take two
+          arguments i,j with an index range of 0..fullmat_dim-1
+        - `permute_order`:
+           if permute_order is false, seeds will be picked sequentially.
+           otherwise randomly
+        """
+
+        if not seedmat_dim < fullmat_dim:
+            raise ValueError("dimension of seed matrix must be smaller "
+                             "than that of full matrix")
+        if not callable(getdist):
+            raise ValueError("distance getter function not callable")
+
+        if permute_order:
+            picked_seeds = sample(range(fullmat_dim), seedmat_dim)
+        else:
+            picked_seeds = range(seedmat_dim)
+        # assert len(picked_seeds) == seedmat_dim, (
+        #    "mismatch between number of picked seeds and seedmat dim.")
+
+        # Putting picked seeds/indices at the front is not enough,
+        # need to change/correct all indices to maintain consistency
+        #
+        used_index_order = range(fullmat_dim)
+        picked_seeds.sort()  # otherwise the below fails
+        for i, seed_idx in enumerate(picked_seeds):
+            used_index_order.pop(seed_idx - i)
+        used_index_order = picked_seeds + used_index_order
+
+        # Order is now determined in used_index_order
+        # first seedmat_dim objects are seeds
+        # now create seedmat
+        #
+        t0 = time.clock()
+        seedmat = zeros((len(picked_seeds), fullmat_dim))
+        for i in range(len(picked_seeds)):
+            for j in range(fullmat_dim):
+                if i < j:
+                    seedmat[i, j] = getdist(used_index_order[i],
+                                            used_index_order[j])
+                elif i == j:
+                    continue
+                else:
+                    seedmat[i, j] = seedmat[j, i]
+
+        restore_idxs = argsort(used_index_order)
+        if PRINT_TIMINGS:
```

Comment:
It might be feasible to glean a lot of the timing information from a `cProfile` run